process/transactionProcessor.go

```diff
@@ -500,24 +500,48 @@ func checkIfMoveBalanceNotarized(tx *transaction.ApiTransactionResult) bool {
 	return true
 }
 
-func (tp *TransactionProcessor) checkIfIntraShardRelayedTransaction(
-	tx *transaction.ApiTransactionResult,
-	senderAddress []byte,
-	receiverAddress []byte,
-) bool {
+func (tp *TransactionProcessor) checkIfCompletelyIntraShardRelayedTransaction(tx *transaction.ApiTransactionResult) (bool, error) {
 	isNotarized := tx.NotarizedAtSourceInMetaNonce > 0 && tx.NotarizedAtDestinationInMetaNonce > 0
 	if !isNotarized {
-		return false
+		return false, nil
 	}
 
 	isRelayedTransaction := tx.ProcessingTypeOnSource == relayedTransactionDescriptor && tx.ProcessingTypeOnDestination == relayedTransactionDescriptor
 	if !isRelayedTransaction {
-		return false
+		return false, nil
+	}
+
+	senderAddress, err := tp.pubKeyConverter.Decode(tx.Sender)
+	if err != nil {
+		return false, err
+	}
+	receiverAddress, err := tp.pubKeyConverter.Decode(tx.Receiver)
+	if err != nil {
+		return false, err
+	}
+
+	dataField := string(tx.Data)
+	if strings.Index(dataField, relayedTxV1DataMarker) != 0 {
+		return false, fmt.Errorf("wrong relayed v1 data marker position")
+	}
+
+	hexedInnerTxData := dataField[len(relayedTxV1DataMarker):]
+	innerTxData, err := hex.DecodeString(hexedInnerTxData)
+	if err != nil {
+		return false, err
+	}
+
+	innerTx := &transaction.Transaction{}
+	err = tp.relayedTxsMarshaller.Unmarshal(innerTx, innerTxData)
+	if err != nil {
+		return false, err
 	}
 
-	isSameShard := tp.proc.GetShardCoordinator().SameShard(senderAddress, receiverAddress)
+	isSameShardOnRelayed := tp.proc.GetShardCoordinator().SameShard(senderAddress, receiverAddress)
```

Comment:
Can also be written as 2 checks / 2 calls of `SameShard()`: `(senderWhichIsRelayer, innerTx.SndAddr)` and `(innerTx.SndAddr, innerTx.RcvAddr)` - but all right to stay as it is, as well, it's more explicit.

Reply:
About relayed V1 that wrap NFT transfers: not-yet-completed transactions of such an exotic kind might reach the "is relayed completely intrashard" check, and can be mistaken as completely intrashard, and thus reach an earlier "done" status - even though they are pending.

Even so, this is an extremely exotic case (no occurrence on mainnet). I think we can skip it for now, what do you think?

Reply:
refactored the solution